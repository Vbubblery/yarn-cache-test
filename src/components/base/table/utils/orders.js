import { frozenPeriods, isFrozen, isMoq, isLotSize } from "./planning";
import { defaultTableValueFormatting, isNotPast } from "./base.js";

export const ORDER_STATUS = {
  VALIDATED: "validated",
  LOCKED: "locked"
};
export const ORDER_FREEZE_STATUS = {
  PARTIAL: "partlyFrozen",
  TOTAL: "fullyFrozen"
};

export const getOrderStatus = status => {
  switch (status) {
    case 2:
      return ORDER_STATUS.VALIDATED;
    case 3:
    case 4:
      return ORDER_STATUS.LOCKED;
    default:
      return undefined;
  }
};

export const getOptionsForOrders = options => {
  const { leadTimes, periods, periodType, productData } = options;
  return {
    periodType,
    periods,
    frozenPeriods: frozenPeriods(
      productData?.leadTime || {},
      periods,
      periodType,
      leadTimes
    ),
    isAggregated: options.isAggregated,
    access: options.access,
    leadTimes: options.leadTimes,
    units: options.productData?.units,
    moq: options.productData?.moq,
    lotSize: options.productData?.lotSize
  };
};

export const getOptionsForStatus3Supplies = options => {
  const { periods, periodType, isAggregated, access } = options;
  return {
    periodType,
    periods,
    isAggregated,
    access,
    units: options.productData?.units
  };
};

export const getOptionsForDemand = options => ({
  periodType: options.periodType,
  periods: options.periods,
  isAggregated: options.isAggregated,
  access: options.access
});

export const getOrdersValues = (orders, options) => {
  if (orders.isParent) {
    return orders.values.map((value, index) => ({
      value,
      isPast: index === 0
    }));
  }

  const {
    isAggregated,
    moq: moqBySupplier,
    lotSize: lotSizeBySupplier,
    ...otherOptions
  } = options;
  const { editedBy, orderStatus, supplier, isParent } = orders;
  return orders.values.map((value, index) => {
    const status = orderStatus && getOrderStatus(orderStatus[index]);
    const isByFlowlity = editedBy && editedBy[index] === "flowlity";
    const frozenStatus = isFrozen(index, supplier, options);
    const moq = isMoq(supplier, isAggregated, orders, moqBySupplier);
    const lotSize = isLotSize(
      supplier,
      isAggregated,
      orders,
      lotSizeBySupplier
    );
    return {
      value,
      isByFlowlity,
      status,
      moq,
      lotSize,
      isPast: index === 0,
      tooltip: supplier !== null || supplier !== undefined,
      isEditable: options.access && !isAggregated && !isParent && index > 0,
      ...(!isNaN(value) && { isFrozen: frozenStatus }),
      ...otherOptions
    };
  });
};

export const getStatus3SuppliesValues = (supplies, options) => {
  const { isAggregated, access, ...otherOptions } = options;
  return supplies.values.map((value, index) => {
    return {
      value,
      isPast: index === 0,
      isEditable: access && !isAggregated && index > 0,
      ...otherOptions
    };
  });
};

export const getAllOrdersValues = (allOrders, options) => {
  return allOrders.values.map((value, index) => {
    return {
      value,
      isPast: !isNotPast(index, options.periods, options.periodType)
    };
  });
};

export const allOrdersCalc = (value, isPast) => {
  if (isPast) return null;
  else if (!value) return 0;
  return defaultTableValueFormatting(value);
};
