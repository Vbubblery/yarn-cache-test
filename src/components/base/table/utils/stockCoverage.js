export const getOptionsForStockCoverage = options => ({
  isAggregated: options.isAggregated,
  stocks: options.tableData.stock.values,
  totalDemandForecast: options.totalDemandForecast
});
export const computeStockCoverage = (stock, totalDemandForecast) => {
  if (!totalDemandForecast) return "âˆž";
  if (!stock || stock <= 0) return 0;
  return Math.trunc(365 / (totalDemandForecast / stock));
};

export const getStockCoverageValues = (stockCoverages, options) => {
  return stockCoverages.values.map((_, index) => ({
    value: options.stocks[index],
    totalDemandForecast: options.totalDemandForecast,
    isPast: index === 0
  }));
};
