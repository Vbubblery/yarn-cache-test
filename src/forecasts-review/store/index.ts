import { computed, SetupContext, ComputedRef } from "@vue/composition-api";
import { PLANNING_DEMAND_VIEW } from "@/forecasts-review/typings";
import { Site } from "@/objects/site";

export const useBaseResources = (context: SetupContext) => {
  const store = context.parent?.$store || null;

  const getMe = computed<{ locale: string }>(
    () => store?.getters["common/getMe"]
  );
  const planningDemandView: ComputedRef<PLANNING_DEMAND_VIEW> = computed(
    () => store?.getters["common/getDefaultView"]
  );
  const currentSite = computed<Site>(
    () => store?.getters["common/getCurrentSite"]
  );
  const tablesData = computed(() => store?.getters["demand/getTablesData"]);
  const demandGetChartData = computed(
    () => store?.getters["demand/getChartData"]
  );
  const demandGetIsLoadingChart = computed(
    () => store?.getters["demand/getIsLoadingChart"]
  );
  const demandGetIsLoadingTables = computed(
    () => store?.getters["demand/getIsLoadingTables"]
  );
  const planningGetChartData = computed(
    () => store?.getters["planning/getChartData"]
  );
  const planningGetIsLoadingChart = computed(
    () => store?.getters["planning/getIsLoadingChart"]
  );
  const planningGetIsLoadingTables = computed(
    () => store?.getters["planning/getIsLoadingTables"]
  );
  const selectedTagIds = computed(
    () => store?.getters["common/getSelectedTags"]
  );
  const selectedCustomerId = computed(() =>
    currentSite.value.isSupplierView
      ? store?.getters["demand/getSelectedCustomerSiteId"]
      : null
  );
  const updateDefaultView = (view: PLANNING_DEMAND_VIEW) => {
    store?.commit("common/updateDefaultView", view);
  };
  const updateSelectedProducts = (items: object[]) => {
    store?.dispatch("common/updateSelectedProducts", items);
  };
  const updateSelectedTags = (items: object[]) => {
    store?.dispatch("common/updateSelectedTags", items);
  };
  const updateDataType = (itemType: string) => {
    store?.commit("common/updateDataType", itemType);
  };
  const resetDemandState = () => {
    store?.commit("demand/resetState");
  };
  const updateDemandPage = (pageNum: number) => {
    store?.commit("demand/updatePage", pageNum);
  };
  const updateDemandTableOnly = (onlyTable: boolean) => {
    store?.commit("demand/updateOnlyTable", onlyTable);
  };
  const fetchData = (
    selectedView: undefined | PLANNING_DEMAND_VIEW = undefined
  ) => {
    const currentView = selectedView ?? planningDemandView.value;
    store?.dispatch(
      currentView === PLANNING_DEMAND_VIEW.DEMAND
        ? "demand/fetchDemandData"
        : "planning/updatePlanningData"
    );
  };
  const updateSelectedCustomerSiteId = (customerSiteId: number | null) => {
    store?.commit("demand/updateSelectedCustomerSiteId", customerSiteId);
  };
  return {
    getMe,
    planningDemandView,
    updateDefaultView,
    currentSite,
    tablesData,
    demandGetChartData,
    demandGetIsLoadingChart,
    demandGetIsLoadingTables,
    planningGetChartData,
    planningGetIsLoadingChart,
    planningGetIsLoadingTables,
    selectedCustomerId,
    selectedTagIds,
    updateSelectedProducts,
    updateSelectedTags,
    updateDataType,
    resetDemandState,
    updateDemandPage,
    updateDemandTableOnly,
    fetchData,
    updateSelectedCustomerSiteId
  };
};
