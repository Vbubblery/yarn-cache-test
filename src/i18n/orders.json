{
  "en": {
    "orderModalAddProductsButton": "Add products",
    "constraintsNotMetMessage": "Some of the products don't meet their constraints",
    "moq": "Minimum order quantity",
    "constraint": "Constraint | Constraint | Constraints",
    "orderConstraints": "Order constraints",
    "orderModalTitle": "Order for",
    "editingModalTitle": "Editing order for",
    "orderTitle": "Orders",
    "planned": "Purchase requisition | Purchase requisition | Purchase requisitions",
    "validated": "Confirmed purchase requisition | Confirmed purchase requisition | Confirmed purchase requisitions",
    "purchaseOrders": "Purchase order | Purchase order | Purchase orders",
    "salesOrders": "Sales order | Sales order | Sales orders",
    "delivered": "Delivered | Delivered | Delivered",
    "search": "Search by order ID, supplier or product",
    "ship_before": "Ship before",
    "send_before": "Send before",
    "date": "Delivery date",
    "totalQuantity": "Total quantity",
    "quantity": "Quantity",
    "addedItems": "Added items",
    "groupedBy": "Grouped by",
    "inTotal": "In total",
    "download_selected": "Download",
    "upload_selected": "Send to API",
    "sftp_selected": "Export to SFTP",
    "sftp_exported": "Exported to SFTP",
    "temporary_id": "Temporary ID",
    "order_id": "Order ID",
    "select_all": "Select all",
    "deselect_all": "Deselect all",
    "deleteSelected": "Delete selected",
    "noOrders": "No orders found",
    "delivery": "Delivery",
    "constraintFrequency": "Doesn't meet order frequency requirements",
    "constraintMOQ": "Doesn't meet MOQ requirements",
    "constraintFullTruck": "Doesn't meet full truck requirements",
    "issueMOQ": "product have issue with MOQ or lot size | products have issue with MOQ or lot size",
    "orderIssues": "Order issues",
    "productIssues": "Product issues",
    "days": "days",
    "latestOrderDate": "Latest Order Date",
    "deliveryDate": "Delivery Date",
    "deselectAll": "Deselect all",
    "suppliers": "Suppliers",
    "reset": "Reset",
    "dateRange": "Date range",
    "thisWeek": "This week",
    "in2Weeks": "In 2 weeks",
    "in1month": "In 1 month",
    "in2months": "In 2 months",
    "in3months": "In 3 months",
    "tempId": "Temporary ID",
    "orderTitleDetail": "Details",
    "orderTitleStatus1": "Purchase requisition",
    "orderTitleStatus2": "Confirmed purchase requisition",
    "orderTitleStatus3": "Purchase order",
    "orderTitleStatus4": "Delivered order",
    "orderDetailMessageTitle": "Messages",
    "orderId": "Order ID",
    "day": "day | day | days",
    "value": "Value",
    "product": "Product",
    "reference": "Reference",
    "lotSize": "Lot Size",
    "leadTime": "Delivery lead time",
    "daysOfCoverage": "Stock coverage at delivery",
    "goodMoq": "Min. order quantity",
    "orderFrequency": "Order frequency",
    "badFullTruckQuantity": "The order is not respecting the full truck quantity constraint",
    "badOrderFrequency": "The order is not respecting the frequency constraint",
    "fullTruckQty": "Full truck quantity",
    "badMOQSupply": "The supply is not respecting the MOQ constraint",
    "badMOQOrder": "The order is not respecting the MOQ constraint",
    "expectedDelivery": "Expected delivery",
    "delete": "Delete",
    "cancel": "Cancel",
    "edit": "edit",
    "cancelEdit": "cancel edit",
    "discardChanges": "Discard Changes",
    "save": "Save",
    "orderIsUpdating": "Order is updating...",
    "deleteSuccess": "Order has been deleted",
    "deleteOrderError": "Cannot delete the order",
    "deleteSupplyError": "The product could not be deleted correctly",
    "deleteWarningTitle": "Delete confirmation",
    "deleteCurrentOrderWarningText": "Are you sure you want to remove this order?",
    "deleteCurrentOrderSupplyWarningText": "This will delete all the planned supplies inside this order.",
    "deleteSupplyWarningText": "Are you sure you want to remove this product ?",
    "deleteWarningConfirmButton": "YES, CONFIRM",
    "deleteWarningCancelButton": "CANCEL",
    "mergeWarningTitle": "Merge confirmation",
    "mergeWarningText": "An order already exists for this date and supplier. Do you want to combine the two orders?\n This will add the current products and quantities to the order found at this new date.",
    "mergeWarningConfirmButton": "YES, CONFIRM",
    "mergeWarningCancelButton": "CANCEL",
    "secondaryUnit": "Secondary unit",
    "orderAndSupplyConstraintNotMet": "Some constraints on the order level and product level are not met",
    "orderConstraintNotMet": "Some constraints on the order level are not met",
    "supplyConstraintNotMet": "Some products don't meet their constraints",
    "customer": {
      "display": "Customer",
      "partner": "Supplier",
      "waitValidateOrder": "Waiting to be validated",
      "orderValidated": "Validated",
      "waitSendOrder": "Waiting to be sent",
      "orderSent": "Sent",
      "sendBefore": "Send before",
      "badLotSize": "Lot size not respected",
      "validateBtn": "Validate"
    },
    "supplier": {
      "display": "Supplier",
      "partner": "Customer",
      "waitValidateOrder": "Waiting to be accepted",
      "orderValidated": "Accepted",
      "waitSendOrder": "Waiting to be shipped",
      "orderSent": "Shipped",
      "leadTimeNotSet": "Delivery lead time is not defined. Set it.",
      "sendBefore": "Ship before",
      "badLotSize": "Lot size not respected",
      "validateBtn": "Accept"
    }
  },
  "fr": {
    "orderModalTitle": "Commande pour",
    "editingModalTitle": "Édition de la commande pour",
    "constraintsNotMetMessage": "Certains produits ne respectent pas leurs contraintes",
    "moq": "Minimum de commande",
    "constraint": "Contrainte | Contrainte | Contraintes",
    "orderConstraints": "Contraintes de commande",
    "orderModalAddProductsButton": "Ajouter des produits",
    "orderTitle": "Commandes",
    "planned": "Prévue | Prévue | Prévues",
    "validated": "Prévalidée | Prévalidée | Prévalidées",
    "purchaseOrders": "Commande d'achat | Commande d'achat | Commandes d'achat",
    "salesOrders": "Commande de vente | Commande de vente | Commandes de vente",
    "delivered": "Livrée | Livrée | Livrées",
    "search": "Chercher par ID, fournisseur ou produit",
    "ship_before": "Expédier le",
    "send_before": "Passer commande avant le",
    "date": "Date de livraison prévue",
    "quantity": "Quantité",
    "totalQuantity": "Quantité totale",
    "addedItems": "Produits ajoutés",
    "groupedBy": "Groupé par",
    "inTotal": "Total",
    "download_selected": "Télécharger",
    "upload_selected": "Envoyer à l'API",
    "sftp_selected": "Exporter VIA SFTP",
    "sftp_exported": "Exportée via SFTP",
    "temporary_id": "ID provisoire",
    "order_id": "ID de la commande",
    "select_all": "Tout sélectionner",
    "deselect_all": "Tout désélectionner",
    "deleteSelected": "Supprimer la sélection",
    "noOrders": "Pas de commande trouvée",
    "delivery": "Livraison le",
    "constraintFrequency": "Ne respecte pas la contrainte de fréquence de commande",
    "constraintMOQ": "Ne respecte pas la contrainte de MOQ",
    "constraintFullTruck": "Ne respecte pas la contrainte de camion plein",
    "issueMOQ": " produit a un problème de MOQ ou de taille de lot | produits ont un problème de MOQ ou de taille de lot",
    "orderIssues": "Lié à la commande",
    "productIssues": "Lié aux produits",
    "days": "jours",
    "latestOrderDate": "Date de commande",
    "deliveryDate": "Date de livraison",
    "deselectAll": "Tout désélectionner",
    "suppliers": "Fournisseurs",
    "reset": "Réinitialiser",
    "dateRange": "Période",
    "thisWeek": "Cette semaine",
    "in2Weeks": "Dans 2 semaines",
    "in1month": "Dans 1 mois",
    "in2months": "Dans 2 mois",
    "in3months": "Dans 3 mois",
    "noTimeRestriction": "Pas de limitation",
    "lastOrderDate": "Date de la commande",
    "tempId": "ID provisoire",
    "orderId": "ID commande",
    "orderTitleDetail": "Détails",
    "orderTitleStatus1": "Commande prévue",
    "orderTitleStatus2": "Commande prévalidée",
    "orderTitleStatus3": "Commande d'achat",
    "orderTitleStatus4": "Commande livrée",
    "orderDetailMessageTitle": "Messages",
    "day": "jour | jour | jours",
    "value": "Valeur",
    "product": "Produit",
    "reference": "Référence",
    "lotSize": "Taille de lot",
    "leadTime": "Délai de livraion",
    "daysOfCoverage": "Couverture de stock à réception",
    "goodMoq": "Minimum de commande",
    "orderFrequency": "Fréquence de commande",
    "badFullTruckQuantity": "La contrainte de camion plein n'est pas respectée",
    "badOrderFrequency": "La contrainte de fréquence de commande n'est pas respectée",
    "fullTruckQty": "Camion plein",
    "badMOQSupply": "La quantité minimum n'est pas respectée",
    "badMOQOrder": "Le minimum de commande n'est pas respecté",
    "expectedDelivery": "Livraison attendue",
    "delete": "Supprimer",
    "cancel": "Annuler",
    "edit": "éditer",
    "cancelEdit": "annuler l'édition",
    "discardChanges": "Annuler les modifications",
    "save": "Enregistrer",
    "orderIsUpdating": "La commande est mise à jour...",
    "deleteSuccess": "Commande supprimée",
    "deleteOrderError": "Suppression impossible",
    "deleteSupplyError": "Une erreur est survenue lors de la suppression du produit",
    "deleteWarningTitle": "Confirmation de suppression",
    "deleteCurrentOrderWarningText": "Voulez-vous supprimer cette commande ?",
    "deleteCurrentOrderSupplyWarningText": "Cela va supprimer tous les approvisionnements liés à cette commande",
    "deleteSupplyWarningText": "Etes vous sûr de vouloir supprimer ce produit ?",
    "deleteWarningConfirmButton": "JE CONFIRME",
    "deleteWarningCancelButton": "ANNULER",
    "mergeWarningTitle": "Confirmation de combinaison",
    "mergeWarningText": "Une commande existe déjà pour cette date et fournisseur. Voulez-vous combiner les deux commandes?\n Ceci aura pour effet d'ajouter les produits et quantités à la commande trouvée à cette nouvelle date.",
    "mergeWarningConfirmButton": "JE CONFIRME",
    "mergeWarningCancelButton": "ANNULER",
    "secondaryUnit": "Unité secondaire",
    "orderAndSupplyConstraintNotMet": "Certaines contraintes au niveau commande et au niveau produit ne sont pas respectées",
    "orderConstraintNotMet": "Certaines contraintes au niveau commande ne sont pas respectées",
    "supplyConstraintNotMet": "Certains produits ne respectent pas leurs contraintes",
    "customer": {
      "display": "Client",
      "partner": "Fournisseur",
      "waitValidateOrder": "En attente de validation",
      "orderValidated": "Validé",
      "waitSendOrder": "En attente d'expédition",
      "sendBefore": "Expédier avant",
      "orderSent": "Expédié",
      "badLotSize": "La taille du lot n'est pas respectée",
      "validateBtn": "Valider"
    },
    "supplier": {
      "display": "Fournisseur",
      "partner": "Client",
      "waitValidateOrder": "En attente de validation",
      "orderValidated": "Accepté",
      "waitSendOrder": "En attente d'expédition",
      "leadTimeNotSet": "Le délai de livraison n'est pas défini. Configurer.",
      "sendBefore": "Expédier avant",
      "orderSent": "Expédié",
      "badLotSize": "La taille du lot n'est pas respectée",
      "validateBtn": "Accepter"
    }
  }
}
