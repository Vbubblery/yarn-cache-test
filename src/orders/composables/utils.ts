export const createDebounce = () => {
  let timeout: number | undefined = undefined;
  return (fnc: Function, delayMs: number) => {
    clearTimeout(timeout);
    timeout = setTimeout(() => {
      fnc();
    }, delayMs || 500);
  };
};
