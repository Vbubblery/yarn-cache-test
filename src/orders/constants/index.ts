export enum ORDER_TYPE {
  NEW_ORDERS = "new_orders",
  PLANNED = "planned",
  VALIDATED = "validated",
  PURCHASE_ORDER = "purchaseOrders",
  SALES_ORDER = "salesOrders",
  DELIVERED = "delivered"
}

export enum ORDER_TYPE_VALUE {
  PLANNED = 1,
  VALIDATED = 2,
  PURCHASE_ORDER = 3,
  SALES_ORDER = 3,
  DELIVERED = 4
}

export enum ORDER_ACTION_TYPE {
  UPLOAD = "upload",
  CONFIRM = "confirm",
  DELETE = "delete",
  EXPORT = "export",
  EXPORT_SFTP = "exportSftp",
  UNCONFIRMED = "unconfirmed",
  VALIDATE = "validate"
}

export enum Views {
  CUSTOMER = "customer",
  SUPPLIER = "supplier"
}

export const events = {
  CHANGE_ORDER_TYPE: "change:order-type"
};
export const supplierOrdersHeaders = [
  { name: ORDER_TYPE.SALES_ORDER, value: ORDER_TYPE_VALUE.SALES_ORDER },
  { name: ORDER_TYPE.DELIVERED, value: ORDER_TYPE_VALUE.DELIVERED }
];

export const userOrdersHeaders = [
  { name: ORDER_TYPE.PLANNED, value: ORDER_TYPE_VALUE.PLANNED },
  { name: ORDER_TYPE.VALIDATED, value: ORDER_TYPE_VALUE.VALIDATED },
  {
    name: ORDER_TYPE.PURCHASE_ORDER,
    value: ORDER_TYPE_VALUE.PURCHASE_ORDER
  },
  { name: ORDER_TYPE.DELIVERED, value: ORDER_TYPE_VALUE.DELIVERED }
];
