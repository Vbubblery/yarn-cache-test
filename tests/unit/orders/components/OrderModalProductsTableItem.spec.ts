import { renderComponent } from "tests/test-utils";
import OrderModalProductsTableItem from "@/orders/components/OrderModalProductsTableItem.vue";
import { Order, OrderSupply } from "@/objects/order";
import { Views } from "@/orders/constants";
import { computeStockCoverage } from "@/components/base/table/utils/stockCoverage";
// eslint-disable-next-line jest/no-mocks-import
import { store } from "@/store/__mocks__/store";
import { ref } from "@vue/composition-api";

/**
 * This component is in charge of:
 * - Showing the alerts about a single supply
 * - Displaying the info about a single supply
 * - Computing reactively the days of coverage for a single supply
 * and delegating the following to its grandparent (OrderModal):
 * - Input event and payload
 * - Click on planning view
 * The events that are delegated should be / are tested in the orchestrator, the grandparent (OrderModal)
 */
describe("OrderModalProductsTableItem", () => {
  it("Renders the component", () => {
    const editableSupplies = ref({});
    const { html } = renderComponent(OrderModalProductsTableItem, {
      props: {
        order: new Order(),
        editableSupplies,
        supply: new OrderSupply(),
        view: Views.CUSTOMER,
        isEditing: false
      },
      store
    });
    expect(html()).toMatchSnapshot();
  });
  it("Displays an alert on the supply level if not respecting constraints - MOQ", async () => {
    const order = new Order({
      id: 1,
      status_id: 1,
      product_site: 1,
      supplies: [{ id: 1, quantity: 1, moq: 2 }]
    });
    const editableSupplies = ref({ 1: 1 });
    const supply = order.supplies[0];
    const { getAllByTestId } = renderComponent(OrderModalProductsTableItem, {
      props: {
        order,
        editableSupplies,
        supply,
        view: Views.CUSTOMER,
        isEditing: false
      },
      store
    });
    expect(getAllByTestId(/order_modal_supply_alert_icon/).length).toEqual(1);
  });
  it("Displays the quantity with red color if its unnecessary at all", async () => {
    const editableSupplies = ref({ 1: 10 });
    const order = new Order({
      id: 1,
      status_id: 1,
      minimum_order_quantity: 2,
      supplies: [
        {
          id: 1,
          product_site: 1,
          quantity: 10,
          minimum_order_quantity: 2,
          lotSize: 5,
          minimum_stock_at_end_of_period: 0,
          stock_at_end_of_period: 11
        }
      ]
    });
    const supply = order.supplies[0];
    const { getByTestId } = renderComponent(OrderModalProductsTableItem, {
      props: {
        order,
        editableSupplies,
        supply,
        view: Views.CUSTOMER,
        isEditing: false
      },
      store
    });
    expect(
      getByTestId(/order_modal_product_input/).getAttribute("class")
    ).toContain("text-red-600");
  });
  it("Displays the quantity with orange color when is partially needed", async () => {
    const editableSupplies = ref({ 1: 10 });
    const order = new Order({
      id: 1,
      status_id: 1,
      minimum_order_quantity: 2,
      supplies: [
        {
          id: 1,
          product_site: 1,
          quantity: 10,
          minimum_order_quantity: 2,
          lotSize: 5,
          minimum_stock_at_end_of_period: 0,
          stock_at_end_of_period: 8
        }
      ]
    });
    const supply = order.supplies[0];
    const { getByTestId } = renderComponent(OrderModalProductsTableItem, {
      props: {
        order,
        editableSupplies,
        supply,
        view: Views.CUSTOMER,
        isEditing: false
      },
      store
    });
    expect(
      getByTestId(/order_modal_product_input/).getAttribute("class")
    ).toContain("text-orange-500");
  });
  it("Displays the quantity with gray color when is fully needed", async () => {
    const editableSupplies = ref({ 1: 10 });
    const order = new Order({
      id: 1,
      status_id: 1,
      supplies: [
        {
          id: 1,
          product_site: 1,
          quantity: 10,
          moq: 0,
          lot_size: 0,
          minimum_stock_at_end_of_period: 0,
          stock_at_end_of_period: 0
        }
      ]
    });
    const supply = order.supplies[0];
    const { getByTestId } = renderComponent(OrderModalProductsTableItem, {
      props: {
        order,
        editableSupplies,
        supply,
        view: Views.CUSTOMER,
        isEditing: false
      },
      store
    });
    expect(
      getByTestId(/order_modal_product_input/).getAttribute("class")
    ).toContain("text-gray-500");
  });
  it("Displays the quantity with gray color when moq is higher than needed stock", async () => {
    const editableSupplies = ref({ 1: 10 });
    const order = new Order({
      id: 1,
      status_id: 1,
      supplies: [
        {
          id: 1,
          product_site: 1,
          quantity: 10,
          moq: 10,
          lot_size: 0,
          minimum_stock_at_end_of_period: 0,
          stock_at_end_of_period: 4
        }
      ]
    });
    const supply = order.supplies[0];
    const { getByTestId } = renderComponent(OrderModalProductsTableItem, {
      props: {
        order,
        editableSupplies,
        supply,
        view: Views.CUSTOMER,
        isEditing: false
      },
      store
    });
    expect(
      getByTestId(/order_modal_product_input/).getAttribute("class")
    ).toContain("text-gray-500");
  });
  it("Displays the quantity with gray color when lotSize is higher than needed stock", async () => {
    const editableSupplies = ref({ 1: 10 });
    const order = new Order({
      id: 1,
      status_id: 1,
      supplies: [
        {
          id: 1,
          product_site: 1,
          quantity: 10,
          moq: 7,
          lot_size: 10,
          minimum_stock_at_end_of_period: 0,
          stock_at_end_of_period: 4
        }
      ]
    });
    const supply = order.supplies[0];
    const { getByTestId } = renderComponent(OrderModalProductsTableItem, {
      props: {
        order,
        editableSupplies,
        supply,
        view: Views.CUSTOMER,
        isEditing: false
      },
      store
    });
    expect(
      getByTestId(/order_modal_product_input/).getAttribute("class")
    ).toContain("text-gray-500");
  });
  it("Displays an alert on the supply level if not respecting constraints - Lot Size", async () => {
    const editableSupplies = ref({ 3: 1 });
    const order = new Order({
      id: 1,
      status_id: 1,
      supplies: [
        {
          id: 3,
          product_site: 3,
          quantity: 1,
          lot_size: 2
        }
      ]
    });
    const supply = order.supplies[0];
    const { getAllByTestId } = renderComponent(OrderModalProductsTableItem, {
      props: {
        order,
        editableSupplies,
        supply,
        view: Views.CUSTOMER,
        isEditing: false
      },
      store
    });
    expect(getAllByTestId(/order_modal_supply_alert_icon/).length).toEqual(1);
  });
  it("Computes and displays the right number of days of coverage - positive stock at delivery", async () => {
    const editableSupplies = ref({ 1: 1 });
    const order = new Order({
      supplies: [
        {
          id: 1,
          stock_at_end_of_period: 1,
          total_demand: 77,
          product_site: 1
        }
      ]
    });
    const supply = order.supplies[0];
    const { getByTestId } = renderComponent(OrderModalProductsTableItem, {
      props: {
        order,
        editableSupplies,
        supply,
        view: Views.CUSTOMER,
        isEditing: false
      },
      store
    });
    expect(
      Number(
        getByTestId(/^order_modal_supply_days_cov/)
          .innerHTML.trim()
          .split(" ")[0]
      )
    ).toEqual(computeStockCoverage(1, 77));
  });
  it("Computes and displays the right number of days of coverage - negative stock at delivery", async () => {
    const editableSupplies = ref({ 2: 2 });
    const order = new Order({
      supplies: [
        { id: 2, stock_at_end_of_period: -1, total_demand: 88, product_site: 2 }
      ]
    });
    const supply = order.supplies[0];
    const { getByTestId } = renderComponent(OrderModalProductsTableItem, {
      props: {
        order,
        editableSupplies,
        supply,
        view: Views.CUSTOMER,
        isEditing: false
      },
      store
    });
    expect(
      Number(
        getByTestId(/^order_modal_supply_days_cov/)
          .innerHTML.trim()
          .split(" ")[0]
      )
    ).toEqual(computeStockCoverage(-1, 88));
  });
  it("Computes and displays the right number of days of coverage - no totalDemand (Inf days of coverage)", async () => {
    const editableSupplies = ref({ 2: 1 });
    const order = new Order({
      supplies: [
        {
          id: 2,
          stock_at_end_of_period: 1,
          total_demand: null,
          product_site: 2
        }
      ]
    });
    const supply = order.supplies[0];
    const { getByTestId } = renderComponent(OrderModalProductsTableItem, {
      props: {
        order,
        editableSupplies,
        supply,
        view: Views.CUSTOMER,
        isEditing: false,
        editableSupplyQuantity: 0
      },
      store
    });
    expect(
      getByTestId(/^order_modal_supply_days_cov/).innerHTML.trim()
    ).toEqual("âˆž");
  });
  it("Computes and displays the right number of days of coverage - no stock at delivery (0 days of coverage)", async () => {
    const editableSupplies = ref({ 2: 1 });
    const order = new Order({
      supplies: [
        {
          id: 2,
          stock_at_end_of_period: undefined,
          total_demand: 88,
          product_site: 2
        }
      ]
    });
    const supply = order.supplies[0];
    const { getByTestId } = renderComponent(OrderModalProductsTableItem, {
      props: {
        order,
        editableSupplies,
        supply,
        view: Views.CUSTOMER,
        isEditing: false
      },
      store
    });
    expect(
      Number(
        getByTestId(/^order_modal_supply_days_cov/)
          .innerHTML.trim()
          .split(" ")[0]
      )
    ).toEqual(0);
  });
  it.todo(
    "should open the Planning when clicking on the product name in regular view"
  );
  it.todo(
    "should not be allowed to open the Planning when clicking on the product name in supplier portal view"
  );
});
