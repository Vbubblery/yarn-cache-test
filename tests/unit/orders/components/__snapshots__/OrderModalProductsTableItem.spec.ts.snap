// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrderModalProductsTableItem Renders the component 1`] = `
<div class="grid grid-cols-24 border-collapse border-t last:border-b border-dashed border-gray-300">
  <div class="col-span-1 flex items-center justify-end py-1">
    <div class="v-popper mr-2 v-popper--theme-tooltip"><svg viewBox="0 0 20 20" width="20" height="20" class="text-orange-500" data-testid="order_modal_supply_alert_icon">
        <use xlink:href="/assets/svg/sprite.svg#svg-error-warning-fill" href="/assets/svg/sprite.svg#svg-error-warning-fill"></use>
      </svg></div>
  </div>
  <div class="text-black py-2 h-full col-span-8">
    <div class="v-popper v-popper--theme-tooltip">
      <p class="truncate">


      </p>
    </div>
  </div>
  <div class="col-span-3 text-center m-0 my-[-1px] border-gray-300 py-2 focus:z-10 focus:border-blue-500 focus-within:border-blue-500 focus-within:z-10">
    <div class="v-popper z-5 v-popper--theme-tooltip">
      <div class="text-right px-2"><span class="text-sm text-gray-900 mr-1"></span> <span></span></div>
    </div>
  </div>
  <div class="h-full flex justify-end items-center col-span-3 text-right py-2 mr-2">
    <div data-testid="order_modal_supply_days_cov_undefined" class="text-black inline-block text-right pr-1">
      ∞
    </div>
    <!---->
  </div>
  <div class="h-full flex justify-end items-center col-span-2 text-right pr-2 py-2"><span class="text-sm text-gray-900 mr-1"></span> <span class="text-2xs text-gray-600"></span></div>
  <div class="h-full flex justify-end items-center col-span-2 text-right pr-2 py-2"><span class="text-sm text-gray-900 mr-1"></span> <span class="text-2xs text-gray-600"></span></div>
  <div class="h-full flex justify-end items-center flex-wrap col-span-3 text-right pr-8 py-2"><span class="text-black">
      NaN
    </span> <span class="text-xs">€</span></div>
  <div data-testid="order_modal_open_planning_button" class="h-full col-span-1 flex justify-center items-center"><button type="button" class="btn btn--action relative h-8 flex justify-center items-center rounded-md cursor-pointer w-8 h-8"><svg viewBox="0 0 20 20" width="20" height="20" class="btn__icon">
        <use xlink:href="/assets/svg/sprite.svg#svg-line-chart-fill" href="/assets/svg/sprite.svg#svg-line-chart-fill"></use>
      </svg>
      <!---->
    </button></div>
  <!---->
</div>
`;
